File: 207_Course_Schedule/canFinish.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype bool
        """
        graph = [[] for _ in xrange(numCourses)]
        # 0: not visit, 1: been visited, -1: being visited
        visit = [0 for _ in xrange(numCourses)] 
        for x, y in prerequisites:
            graph[x].append(y)
        
        def dfs(i):
            print visit, i
            if visit[i] == -1:
                return False
            if visit[i] == 1:
                return True
            visit[i] = -1
            for j in graph[i]:
                if not dfs(j):
                    return False
            visit[i] = 1
            return True
        
        for i in xrange(numCourses):
            if not dfs(i):
                return False
        return True


sol = Solution()
numCourses = 2
prerequisites = [[1, 0], [0, 1]]
print sol.canFinish(numCourses, prerequisites)

File: 207_Course_Schedule/canFinish.cpp
#include <iostream>
#include <unordered_set>
#include <vector>
#include <utility>

using namespace std;

class Solution {
    public:
        bool canFinish(int numCourses, vector<pair<int, int> >& prerequisites) {
            vector<unordered_set<int> > graph = makeGraph(numCourses, prerequisites);
            vector<int> indegrees = computeIndegree(graph);
            for(int i = 0; i < numCourses; i++) {
                int j = 0;
                for(; j < numCourses; j++) {
                    if(!indegrees[j]) { break; }
                }
                if(j == numCourses) { return false; }
                indegrees[j] = -1;
                for(int neigh : graph[j]) {
                    indegrees[neigh] -= 1;
                }
            }

            return true;
        }

    private:
        vector<unordered_set<int> > makeGraph(int numCourses, vector<pair<int, int> >& prerequisites) {
            vector<unordered_set<int> > graph(numCourses);

            for(auto pre : prerequisites) {
                // [pre] = {courses}
                graph[pre.second].insert(pre.first);
            }
            return graph;
        }

        vector<int> computeIndegree(vector<unordered_set<int> >& graph) {
            vector<int> degrees(graph.size(), 0);

            for(auto neighbors : graph) {
                for(int neigh : neighbors) {
                    degrees[neigh]++;
                }
            }

            return degrees;
        }

};
